File: EntityFX.NetBenchamarks.Core/Generic/MemoryBenchmarkBase.cs
using System;
using System.Linq;
using System.Diagnostics;

namespace EntityFX.NetBenchmark.Core.Generic
{
    public abstract class MemoryBenchmarkBase<TResult> : BenchmarkBase<TResult>
    {
        private Random random;

        Writer output = new Writer();

        protected bool UseConsole { get => output.UseConsole; set => output.UseConsole = value; }

        public MemoryBenchmarkBase(bool printToConsole = true)
        {
            Iterrations = 500000;
            Ratio = 1;
            random = new Random((int)DateTime.Now.Ticks);
            UseConsole = printToConsole;
        }

        public override void Warmup(double aspect = 0.05)
        {
            UseConsole = false;
            base.Warmup(aspect);
            UseConsole = true;
        }

        public MemoryBenchmarkResult BenchRandomMemory()
        {
            var int4k = MeasureArrayRandomRead(1024);
            output.WriteLine($"int 4k: {int4k.Item1.ToString("F2")} MB/s");
            var int512k = MeasureArrayRandomRead(131072);
            output.WriteLine($"int 512k: {int512k.Item1.ToString("F2")} MB/s");
            var int8m = MeasureArrayRandomRead(2097152);
            output.WriteLine($"int 8M: {int8m.Item1.ToString("F2")} MB/s");
            var int32m = MeasureArrayRandomRead(32 * 1024 * 1024 / sizeof(int));
            output.WriteLine($"int 32M: {int8m.Item1.ToString("F2")} MB/s");

            var long4k = MeasureArrayRandomLongRead(512);
            output.WriteLine($"long 4k: {long4k.Item1.ToString("F2")} MB/s");
            var long512k = MeasureArrayRandomLongRead(65536);
            output.WriteLine($"long 512k: {long512k.Item1.ToString("F2")} MB/s");
            var long8m = MeasureArrayRandomLongRead(1048576);
            output.WriteLine($"long 8M: {long8m.Item1.ToString("F2")} MB/s");
            var long32m = MeasureArrayRandomRead(32 * 1024 * 1024 / sizeof(long));
            output.WriteLine($"long 32M: {int8m.Item1.ToString("F2")} MB/s");

            var avg = (new double[] { int4k.Item1, int512k.Item1, int8m.Item1, long4k.Item1, long512k.Item1, long8m.Item1 }).Average();
            output.WriteLine($"Average: {avg.ToString("F2")} MB/s");

            return new MemoryBenchmarkResult()
            {
                Average = avg,
                Output = output.Output
            };
        }

        protected Tuple<double, int[]> MeasureArrayRandomRead(int size)
        {
            int blockSize = 16;
            int[] I = new int[blockSize];
            var sw = new Stopwatch();
            var array = Enumerable.Range(0, size).Select(r => random.Next(int.MinValue, int.MaxValue)).ToArray();
            var end = array.Length - 1;

            var k1 = (size / 1024) == 0 ? 1 : (size / 1024) ;
            var iterInternal = Iterrations / k1;           

            for (int idx = 0; idx < end; idx+=blockSize)
            {
                //System.Buffer.BlockCopy(array, idx, I, 0, blockSizeBytes);
                I[0] = array[idx];
                I[1] = array[idx+1];
                I[2] = array[idx+2];
                I[3] = array[idx+3];
                I[4] = array[idx+4];
                I[5] = array[idx+5];
                I[6] = array[idx+6];
                I[7] = array[idx+7];
                I[8] = array[idx+8];
                I[9] = array[idx+9];
                I[0xA] = array[idx+0xA];
                I[0xB] = array[idx+0xB];
                I[0xC] = array[idx+0xC];
                I[0xD] = array[idx+0xD];
                I[0xE] = array[idx+0xE];
                I[0xF] = array[idx+0xF];
            }
            sw.Start();
            for (int i = 0; i < iterInternal; i++)
            {
                for (int idx = 0; idx < end; idx+=blockSize)
                {
                    //System.Buffer.BlockCopy(array, idx, I, 0, blockSizeBytes);
                    I[0] = array[idx];
                    I[1] = array[idx+1];
                    I[2] = array[idx+2];
                    I[3] = array[idx+3];
                    I[4] = array[idx+4];
                    I[5] = array[idx+5];
                    I[6] = array[idx+6];
                    I[7] = array[idx+7];
                    I[8] = array[idx+8];
                    I[9] = array[idx+9];
                    I[0xA] = array[idx+0xA];
                    I[0xB] = array[idx+0xB];
                    I[0xC] = array[idx+0xC];
                    I[0xD] = array[idx+0xD];
                    I[0xE] = array[idx+0xE];
                    I[0xF] = array[idx+0xF];
                }
            }
            sw.Stop();

            return new Tuple<double, int[]>((long)iterInternal * array.Length * 4 / sw.Elapsed.TotalSeconds / 1024 / 1024, I);
        }

        protected Tuple<double, long[]> MeasureArrayRandomLongRead(int size)
        {
            int blockSize = 8;
            long[] L = new long[blockSize];
            var sw = new Stopwatch();
            var array = Enumerable.Range(0, size).Select(r => (long)random.Next(int.MinValue, int.MaxValue)).ToArray();
            var end = array.Length - 1;
            var k1 = (size / 1024) == 0 ? 1 : (size / 1024) ;
            var iterInternal = Iterrations / k1;
            for (int idx = 0; idx < end; idx+=blockSize)
            {
                L[0] = array[idx];
                L[1] = array[idx+1];
                L[2] = array[idx+2];
                L[3] = array[idx+3];
                L[4] = array[idx+4];
                L[5] = array[idx+5];
                L[6] = array[idx+6];
                L[7] = array[idx+7];
            }
            sw.Start();
            for (int i = 0; i < iterInternal; i++)
            {
                for (int idx = 0; idx < end; idx+=blockSize)
                {
                    //System.Buffer.BlockCopy(array, idx, I, 0, blockSizeBytes);
                    L[0] = array[idx];
                    L[1] = array[idx+1];
                    L[2] = array[idx+2];
                    L[3] = array[idx+3];
                    L[4] = array[idx+4];
                    L[5] = array[idx+5];
                    L[6] = array[idx+6];
                    L[7] = array[idx+7];
                }
            }
            sw.Stop();

            return new Tuple<double, long[]>((long)iterInternal * array.Length * 8 / sw.Elapsed.TotalSeconds / 1024 / 1024, L);
        }
    }
}

File: src/python/entityfx/memory_benchmark_base.py

from entityfx.benchmark_base import BenchmarkBase
from entityfx.writer import Writer
import random
import math
import time
from statistics import mean

class MemoryBenchmarkBase(BenchmarkBase):
    
    
    def __init__(self, writer: Writer, print_to_console : bool=True) -> None:
        super().__init__(writer, print_to_console)

        random.seed()
        
        self.__output = Writer()
        self._iterrations = 500000
        self.ratio = (1)
        self._use_console(print_to_console)
    
    def warmup(self, aspect : float=.05) -> None:
        self._use_console(False)
        super().warmup(aspect)
        self._use_console(True)
    
    def benchRandomMemory(self) :
        int4k = self._measureArrayRandomRead(1024)
        self._output.write_line("int 4k: {0:5.2f} MB/s", int4k[0])
        int512k = self._measureArrayRandomRead(131072)
        self._output.write_line("int 512k: {0:5.2f} MB/s", int512k[0])
        int8m = self._measureArrayRandomRead(2097152)
        self._output.write_line("int 8m: {0:5.2f} MB/s", int8m[0])
        int32m = self._measureArrayRandomRead(math.floor(32 * 1024 * 1024 / 4))
        self._output.write_line("int 32m: {0:5.2f} MB/s", int32m[0])

        long4k = self._measureArrayRandomLongRead(512)
        self._output.write_line("long 4k: {0:5.2f} MB/s", long4k[0])
        long512k = self._measureArrayRandomLongRead(65536)
        self._output.write_line("long 512k: {0:5.2f} MB/s", long512k[0])
        long8m = self._measureArrayRandomLongRead(1048576)
        self._output.write_line("long 8m: {0:5.2f} MB/s", long8m[0])
        long32m = self._measureArrayRandomRead(math.floor(32 * 1024 * 1024 / 4))
        self._output.write_line("long 32m: {0:5.2f} MB/s", long32m[0])

        avg = mean([int4k[0], int512k[0], int8m[0], int32m[0], long4k[0], long512k[0], long8m[0], long32m[0]])
        
        return { "Average" : avg, "Output" : self.__output.output }
    
    def _measureArrayRandomRead(self, size) :
        block_size = 16
        i = [0] * block_size

        array0_ = list(range(0, size))
        end = len(array0_) - 1
        k0 = math.floor(size / 1024)
        k1 = 1 if k0 == 0 else k0

        iter_internal = math.floor(self._iterrations / k1)
        iter_internal = 1 if iter_internal == 0 else iter_internal

        idx = 0
        while idx < end: 
            i[0] = (array0_[idx])
            i[1] = (array0_[idx + 1])
            i[2] = (array0_[idx + 2])
            i[3] = (array0_[idx + 3])
            i[4] = (array0_[idx + 4])
            i[5] = (array0_[idx + 5])
            i[6] = (array0_[idx + 6])
            i[7] = (array0_[idx + 7])
            i[8] = (array0_[idx + 8])
            i[9] = (array0_[idx + 9])
            i[0xA] = (array0_[idx + 0xA])
            i[0xB] = (array0_[idx + 0xB])
            i[0xC] = (array0_[idx + 0xC])
            i[0xD] = (array0_[idx + 0xD])
            i[0xE] = (array0_[idx + 0xE])
            i[0xF] = (array0_[idx + 0xF])
            idx += block_size

        start = time.time()
        it = 0

        while it < iter_internal: 
            idx = 0
            while idx < end: 
                i[0] = (array0_[idx])
                i[1] = (array0_[idx + 1])
                i[2] = (array0_[idx + 2])
                i[3] = (array0_[idx + 3])
                i[4] = (array0_[idx + 4])
                i[5] = (array0_[idx + 5])
                i[6] = (array0_[idx + 6])
                i[7] = (array0_[idx + 7])
                i[8] = (array0_[idx + 8])
                i[9] = (array0_[idx + 9])
                i[0xA] = (array0_[idx + 0xA])
                i[0xB] = (array0_[idx + 0xB])
                i[0xC] = (array0_[idx + 0xC])
                i[0xD] = (array0_[idx + 0xD])
                i[0xE] = (array0_[idx + 0xE])
                i[0xF] = (array0_[idx + 0xF])
                idx += block_size
            it += 1
        
        elapsed = time.time() - start
        return (iter_internal * len(array0_) * 4 / elapsed / 1024 / 1024, i)
    
    def _measureArrayRandomLongRead(self, size) :
        block_size = 8
        l_ = [0] * block_size

        array0_ = list(range(0, size))
        end = len(array0_) - 1
        k0 = math.floor(size / 1024)
        k1 = 1 if k0 == 0 else k0

        iter_internal = math.floor(self._iterrations / k1)
        iter_internal = 1 if iter_internal == 0 else iter_internal

        idx = 0

        while idx < end: 
            l_[0] = (array0_[idx])
            l_[1] = (array0_[idx + 1])
            l_[2] = (array0_[idx + 2])
            l_[3] = (array0_[idx + 3])
            l_[4] = (array0_[idx + 4])
            l_[5] = (array0_[idx + 5])
            l_[6] = (array0_[idx + 6])
            l_[7] = (array0_[idx + 7])
            idx += block_size

        start = time.time()
        it = 0

        while it < iter_internal: 
            idx = 0
            while idx < end: 
                l_[0] = (array0_[idx])
                l_[1] = (array0_[idx + 1])
                l_[2] = (array0_[idx + 2])
                l_[3] = (array0_[idx + 3])
                l_[4] = (array0_[idx + 4])
                l_[5] = (array0_[idx + 5])
                l_[6] = (array0_[idx + 6])
                l_[7] = (array0_[idx + 7])
                idx += block_size
            it += 1

        elapsed = time.time() - start
        return (iter_internal * len(array0_) * 4 / elapsed / 1024 / 1024, l_)